File: src/main/java/page/anim/DIYViewPage.java
package page.anim;

import common.system.VImg;
import common.util.AnimGroup;
import common.util.anim.AnimCE;
import main.MainBCU;
import main.Opts;
import page.JBTN;
import page.JTG;
import page.MainLocale;
import page.Page;
import page.awt.BBBuilder;
import page.support.AnimTreeRenderer;
import page.support.AnimTreeTransfer;
import page.support.TreeNodeExpander;
import page.view.AbViewPage;
import utilpc.UtilPC;

import javax.swing.*;
import javax.swing.tree.DefaultMutableTreeNode;
import javax.swing.tree.DefaultTreeModel;
import javax.swing.tree.TreePath;
import java.awt.*;
import java.awt.event.KeyEvent;
import java.util.ArrayList;

public class DIYViewPage extends AbViewPage implements AbEditPage {

	private static final long serialVersionUID = 1L;

	private static final String[] icos = new String[] { "default", "starred", "soul", "EF", "TF", "uni_f", "uni_c", "uni_s" };

	private final JTree jlt = new JTree();
	private final JScrollPane jspu = new JScrollPane(jlt);
	private final JBTN group = new JBTN(MainLocale.PAGE, "addgroup");
	private final JBTN remgroup = new JBTN(MainLocale.PAGE, "remgroup");
	private final EditHead aep;
	private final AnimGroupTree agt;

	private final JTG ics = new JTG(0, "icon");
	private final JBTN icc = new JBTN(0, "confirm");
	private final JComboBox<String> jcb = new JComboBox<>(icos);
	private final JLabel uni = new JLabel();
	private final IconBox ib;

	public DIYViewPage(Page p) {
		super(p, BBBuilder.def.getIconBox());
		ib = (IconBox) vb;
		aep = new EditHead(this, 0);
		agt = new AnimGroupTree(jlt);
		ini();
		resized();
	}

	public DIYViewPage(Page p, AnimCE ac) {
		super(p, BBBuilder.def.getIconBox());
		ib = (IconBox) vb;
		aep = new EditHead(this, 0);
		agt = new AnimGroupTree(jlt);
		if (!ac.inPool())
			aep.focus = ac;
		ini();
		resized();
	}

	public DIYViewPage(Page p, EditHead bar) {
		super(p, BBBuilder.def.getIconBox());
		ib = (IconBox) vb;
		aep = bar;
		agt = new AnimGroupTree(jlt);
		ini();
		resized();
	}

	@Override
	protected JButton getBackButton() {
		return super.getBackButton();
	}

	@Override
	public void setSelection(AnimCE ac) {

		DefaultMutableTreeNode selectedNode = agt.findAnimNode(ac, null);

		if(selectedNode == null)
			return;

		agt.expandCurrentAnimNode(selectedNode);
		jlt.setSelectionPath(new TreePath(selectedNode.getPath()));

		remgroup.setEnabled(false);
	}

	@Override
	protected void enabler(boolean b) {
		super.enabler(b);
		ics.setEnabled(ics.isSelected() || b && pause);
		icc.setEnabled(ics.isSelected());
		jlt.setEnabled(b);
	}

	@Override
	protected void keyPressed(KeyEvent ke) {
		super.keyPressed(ke);

		if(ke.getSource() == jlt) {
			if(ke.getKeyCode() == KeyEvent.VK_DELETE) {
				if(remgroup.isEnabled()) {
					TreePath[] paths = jlt.getSelectionPaths();

					if(paths == null)
						return;

					if(Opts.conf(get(MainLocale.PAGE, "remgroupconf"))) {
						ArrayList<String> groups = new ArrayList<>();

						for(TreePath path : paths) {
							if(path == null || !(path.getLastPathComponent() instanceof DefaultMutableTreeNode) || !(((DefaultMutableTreeNode) path.getLastPathComponent()).getUserObject() instanceof String))
								return;

							if(((DefaultMutableTreeNode) path.getLastPathComponent()).isRoot())
								return;

							String groupName = (String) ((DefaultMutableTreeNode) path.getLastPathComponent()).getUserObject();

							if(groupName.equals(""))
								return;

							groups.add(groupName);
						}

						for(String groupName : groups) {
							agt.removeGroup(groupName);
						}

						remgroup.setEnabled(false);
					}
				}
			}
		}
	}

	@Override
	protected void renew() {
		if (aep.focus == null) {
			AnimGroup.workspaceGroup.renewGroup();
			agt.renewNodes();
		}  else {
			DefaultMutableTreeNode root = new DefaultMutableTreeNode("Animation");

			root.add(new DefaultMutableTreeNode(aep.focus));

			jlt.setModel(new DefaultTreeModel(root));
		}

		DefaultMutableTreeNode firstNode = agt.getVeryFirstAnimNode();

		if(firstNode != null) {
			agt.expandCurrentAnimNode(firstNode);
			jlt.setSelectionPath(new TreePath(firstNode));
		}

		group.setEnabled(aep.focus == null);
	}

	@Override
	protected void resized(int x, int y) {
		super.resized(x, y);
		set(camres, x ,y, 400, 0, 200, 50);
		set(copy, x, y, 200, 0, 200, 50);
		set(larges, x, y , 600, 0, 200, 50);
		set(aep, x, y, 800, 0, 1500, 50);
		if (!larges.isSelected()) {
			set(jspu, x, y, 50, 100, 300, 1050);
			set(ics, x, y, 1000, 1050, 200, 50);
			set(uni, x, y, 750, 500, 200, 200);
			set(jcb, x, y, 750, 750, 200, 50);
			set(icc, x, y, 1000, 1150, 200, 50);
			set(group, x, y, 50, 1150, 300, 50);
			set(remgroup, x, y, 50, 1200, 300, 50);
		} else {
			set(uni, x, y, 200, 800, 200, 150);
			set(jcb, x, y, 150, 950, 200, 50);
			set(ics, x, y, 50, 1200, 200, 50);
			set(icc, x, y, 250, 1200, 200, 50);
			set(jspu, x, y, 50, 1000, 400, 200);
			set(group, x, y, 0, 0, 0, 0);
			set(remgroup, x, y, 0, 0, 0, 0);
		}
		SwingUtilities.invokeLater(() -> jlt.setUI(new TreeNodeExpander(jlt)));
		ib.updateControllerDimension(((Component) vb).getWidth(), ((Component) vb).getHeight());
	}

	@Override
	protected void updateChoice() {
		TreePath path = jlt.getSelectionPath();

		if(path == null)
			return;

		Object o = path.getLastPathComponent();

		if(!(o instanceof DefaultMutableTreeNode))
			return;

		if(((DefaultMutableTreeNode) o).getUserObject() instanceof AnimCE) {
			AnimCE e = (AnimCE) ((DefaultMutableTreeNode) o).getUserObject();
			aep.setAnim(e);
			uni.setIcon(e == null ? null : UtilPC.getIcon(e.getUni()));
			if (e == null)
				return;
			setAnim(e);
		}
	}

	private void addListeners() {

		jlt.addTreeSelectionListener(t -> {
			TreePath[] paths = jlt.getSelectionPaths();

			if(paths == null)
				return;

			boolean canEnabled = true;

			for(TreePath path : paths) {
				if(!(path.getLastPathComponent() instanceof DefaultMutableTreeNode))
					return;

				if(((DefaultMutableTreeNode) path.getLastPathComponent()).isRoot()) {
					canEnabled = false;
					break;
				}

				if(!(((DefaultMutableTreeNode) path.getLastPathComponent()).getUserObject() instanceof String)) {
					canEnabled = false;
					break;
				}
			}

			remgroup.setEnabled(canEnabled);

			updateChoice();
		});

		remgroup.setLnr(a -> {
			TreePath[] paths = jlt.getSelectionPaths();

			if(paths == null)
				return;

			if(Opts.conf(get(MainLocale.PAGE, "remgroupconf"))) {
				ArrayList<String> groups = new ArrayList<>();

				for(TreePath path : paths) {
					if(path == null || !(path.getLastPathComponent() instanceof DefaultMutableTreeNode) || !(((DefaultMutableTreeNode) path.getLastPathComponent()).getUserObject() instanceof String))
						return;

					if(((DefaultMutableTreeNode) path.getLastPathComponent()).isRoot())
						return;

					String groupName = (String) ((DefaultMutableTreeNode) path.getLastPathComponent()).getUserObject();

					if(groupName.equals(""))
						return;

					groups.add(groupName);
				}

				for(String groupName : groups) {
					agt.removeGroup(groupName);
				}

				remgroup.setEnabled(false);
			}
		});

		ics.addActionListener(arg0 -> {
			enabler(!ics.isSelected());
			ib.setBlank(ics.isSelected());
		});

		jcb.addActionListener(arg0 -> {
			int t = jcb.getSelectedIndex();
			IconBox.IBConf.mode = t / 5; // [ Def, Str, Sou,
			IconBox.IBConf.type = t % 5; // [ 0,   1,   2,  3, 4, 5, 6, 7]
			IconBox.IBConf.glow = t / 3 >= 1 && t % 3 <= 1 ? 1 : 0;
			ib.changeType();
		});

		icc.setLnr(x -> {
			VImg clip = MainBCU.builder.toVImg(ib.getClip());
			if (IconBox.IBConf.mode == 0
					&& Opts.conf("are you sure to replace display icon? This action cannot be undone")) {
				AnimCE ac = aep.anim;
				ac.setEdi(clip);
				ac.saveIcon();
				jlt.repaint();
			}
			if (IconBox.IBConf.mode == 1
					&& Opts.conf("are you sure to replace battle icon? This action cannot be undone")) {
				AnimCE ac = aep.anim;
				ac.setUni(MainBCU.builder.toVImg(ib.getClip()));
				ac.saveUni();
				uni.setIcon(UtilPC.getIcon(ac.getUni()));
			}
		});

		group.setLnr(a -> {
			String groupName = Opts.read(get(MainLocale.PAGE, "addgrouppop"));

			if(groupName == null)
				return;

			groupName = AnimGroup.workspaceGroup.validateGroupName(groupName);

			AnimGroup.workspaceGroup.groups.put(groupName, new ArrayList<>());

			renew();
		});
	}

	private void ini() {
		preini();
		add(aep);
		add(jspu);
		add(ics);
		add(icc);
		add(jcb);
		add(uni);
		add(group);
		add(remgroup);
		jcb.setSelectedIndex(IconBox.IBConf.type);
		ics.setEnabled(false);
		icc.setEnabled(false);
		jlt.setCellRenderer(new AnimTreeRenderer());
		group.setEnabled(aep.focus == null);
		SwingUtilities.invokeLater(() -> jlt.setUI(new TreeNodeExpander(jlt)));
		jlt.setTransferHandler(new AnimTreeTransfer(agt));
		jlt.setDragEnabled(true);
		jlt.addTreeExpansionListener(agt);
		jlt.setDropMode(DropMode.ON_OR_INSERT);
		remgroup.setEnabled(false);
		addListeners();

	}

}

File: src/page/anim/DIYViewPage.kt
package page.anim

import common.system.VImg
import common.util.anim.AnimCE
import common.util.anim.AnimU.UType
import main.Opts
import page.JBTN
import page.JTG
import page.Page
import page.anim.IconBox.IBConf
import page.awt.BBBuilder
import page.support.AnimLCR
import page.view.AbViewPage
import utilpc.UtilPC
import java.awt.event.ActionEvent
import java.awt.event.ActionListener
import java.awt.event.KeyEvent
import java.util.*
import java.util.function.Consumer
import javax.swing.JComboBox
import javax.swing.JLabel
import javax.swing.JList
import javax.swing.JScrollPane
import javax.swing.event.ListSelectionEvent
import javax.swing.event.ListSelectionListener

class DIYViewPage : AbViewPage, AbEditPage {
    private val jlu: JList<AnimCE> = JList<AnimCE>()
    private val jspu: JScrollPane = JScrollPane(jlu)
    private val aep: EditHead
    private val ics: JTG = JTG(0, "icon")
    private val icc: JBTN = JBTN(0, "confirm")
    private val jcb: JComboBox<String> = JComboBox<String>(icos)
    private val uni = JLabel()
    private val ib: IconBox

    constructor(p: Page?) : super(p, BBBuilder.Companion.def.getIconBox()) {
        ib = vb as IconBox
        aep = EditHead(this, 0)
        ini()
        resized()
    }

    constructor(p: Page?, ac: AnimCE) : super(p, BBBuilder.Companion.def.getIconBox()) {
        ib = vb as IconBox
        aep = EditHead(this, 0)
        if (!ac.inPool()) aep.focus = ac
        ini()
        resized()
    }

    constructor(p: Page?, bar: EditHead) : super(p, BBBuilder.Companion.def.getIconBox()) {
        ib = vb as IconBox
        aep = bar
        ini()
        resized()
    }

    override fun setSelection(ac: AnimCE?) {
        jlu.setSelectedValue(ac, true)
    }

    override fun enabler(b: Boolean) {
        super.enabler(b)
        ics.isEnabled = ics.isSelected || b && pause
        icc.isEnabled = ics.isSelected
        jlu.isEnabled = b
    }

    protected override fun keyPressed(ke: KeyEvent) {
        if (ke.source === ib) ib.keyPressed(ke)
    }

    protected override fun keyReleased(ke: KeyEvent) {
        if (ke.source === ib) ib.keyReleased(ke)
    }

    protected override fun renew() {
        if (aep.focus == null) jlu.setListData(Vector<AnimCE>(AnimCE.Companion.map().values)) else jlu.setListData(arrayOf<AnimCE?>(aep.focus))
        jlu.selectedIndex = 0
    }

    override fun resized(x: Int, y: Int) {
        super.resized(x, y)
        Page.Companion.set(aep, x, y, 550, 0, 1750, 50)
        Page.Companion.set(jspu, x, y, 50, 100, 300, 1100)
        Page.Companion.set(ics, x, y, 1000, 1050, 200, 50)
        Page.Companion.set(uni, x, y, 750, 500, 200, 200)
        Page.Companion.set(jcb, x, y, 750, 750, 200, 50)
        Page.Companion.set(icc, x, y, 1000, 1150, 200, 50)
    }

    override fun updateChoice() {
        val e: AnimCE = jlu.selectedValue
        aep.setAnim(e)
        uni.icon = if (e == null) null else UtilPC.getIcon(e.getUni())
        if (e == null) return
        setAnim<UType>(e)
    }

    private fun addListeners() {
        jlu.addListSelectionListener(object : ListSelectionListener {
            override fun valueChanged(arg0: ListSelectionEvent) {
                if (arg0.valueIsAdjusting) return
                updateChoice()
            }
        })
        ics.addActionListener(object : ActionListener {
            override fun actionPerformed(arg0: ActionEvent?) {
                enabler(!ics.isSelected)
                ib.setBlank(ics.isSelected)
            }
        })
        jcb.addActionListener(object : ActionListener {
            override fun actionPerformed(arg0: ActionEvent?) {
                val t: Int = jcb.selectedIndex
                IBConf.mode = t / 4
                IBConf.type = t % 4
                IBConf.glow = if (IBConf.type + IBConf.mode > 1) 1 else 0
                ib.changeType()
            }
        })
        icc.setLnr(Consumer { x: ActionEvent? ->
            val clip = VImg(ib.getClip())
            if (IBConf.mode == 0
                    && Opts.conf("are you sure to replace display icon? This action cannot be undone")) {
                val ac: AnimCE = aep.anim
                ac.setEdi(clip)
                ac.saveIcon()
                jlu.repaint()
            }
            if (IBConf.mode == 1
                    && Opts.conf("are you sure to replace battle icon? This action cannot be undone")) {
                val ac: AnimCE = aep.anim
                ac.setUni(VImg(ib.getClip()))
                ac.saveUni()
                uni.icon = UtilPC.getIcon(ac.getUni())
            }
        })
    }

    private fun ini() {
        preini()
        add(aep)
        add(jspu)
        add(ics)
        add(icc)
        add(jcb)
        add(uni)
        jcb.selectedIndex = IBConf.type
        ics.isEnabled = false
        icc.isEnabled = false
        jlu.setCellRenderer(AnimLCR())
        addListeners()
    }

    companion object {
        private const val serialVersionUID = 1L
        private val icos = arrayOf("default", "starred", "EF", "TF", "uni_f", "uni_c", "uni_s")
    }
}
